---
title: "A tissue-specific collaborative mixed model for jointly analyzing multiple tissues in transcriptome-wide association studies"
author: "Xingjie Shi"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

**TisCoMM** package provides a unified probabilistic model for TWAS, leveraging the co-regulation of genetic variations across different tissues explicitly. **TisCoMM** not only performs hypothesis testing to prioritize gene-trait associations,
 but also detects the tissue-specific role of candidate target genes in complex traits. To make use of widely available GWAS summary statistics, TisCoMM is extended to use summary-level data, namely, TisCoMM-S$^2$ . 

## Installation
To install the development version of **TisCoMM**, it's easiest to use the 'devtools' package. Note that **TisCoMM** depends on the 'Rcpp' package, which also requires appropriate setting of Rtools and Xcode for Windows and Mac OS/X, respectively.

```{r, fig.show='hold', eval=FALSE}
library(devtools)
install_github("XingjieShi/TisCoMM")
```

# Real Data Analysis
## Preparing eQTL data
The eQTL data consists of matched genotype data and gene expression data in multiple target tissues. 
\begin{enumerate}
    \item Genotype data of the eQTL samples in the PLINK binary format (.\textbf{bed}), and must be accompanied by .\textbf{bim} and .\textbf{fam} files with the same prefix. For example:
    \begin{itemize}
        \item GTEx.bed,
        \item GTEx.bim,
        \item GTEx.fam.
    \end{itemize}
  
    
    \item Expression data for all target tissues. Please note that gene expression in each tissue should be normaized to remove all confounding effects. Each tissue file should be a \textbf{tab-delimited} text file and include following information for each gene:
    \begin{itemize}
        \item Start location
        \item End location
        \item Gene type
        \item Hugo name
        \item Ensembl ID
        \item Chromosome number
        \item normalized expression levels cross samples
    \end{itemize}
    If some gene annotation is not included in the original gene expression file, one will have to extract these information by performing gene ID mapping with other annotation files. GENCODE (\href{https://www.gencodegenes.org/human/}{https://www.gencodegenes.org/human/}) provides comprehensive gene annotation files.	
\end{enumerate}
 
 \textit{TisCoMM} will use the headers in the expression files to extract information. It is required to have specific columns in all the formatted expression file. See Table \ref{tab:ge} for a demonstration. Note that the first six column names should be exactly the same as those in Table \ref{tab:ge}. Expression levels across individuals should be appended after the first six columns.
\begin{table}[ht]
\centering
\caption{The first three rows and eight columns in an example gene expression file (rows for genes, and columns after the first six columns for samples).}
\label{tab:ge}
\begin{tabular}{ccccccccc}
  \hline
  lower & up & genetype1 & genetype2 & TargetID & Chr & ID1 & ID2 & $\cdots$ \\
  \hline
  59783540 & 59843484 & lincRNA & PART1 & ENSG00000152931.6 &   5 & 0.51 & 0.71 & $\cdots$ \\ 
   48128225 & 48148330 & protein\_coding & UPP1 & ENSG00000183696.9 &   7 & 1.41 & -0.01 & $\cdots$ \\ 
    57846106 & 57853063 & protein\_coding & INHBE & ENSG00000139269.2 &  12 & 0.58 & -1.02 & $\cdots$ \\ 
    \vdots & \vdots & \vdots & \vdots & \vdots &  \vdots & \vdots& \vdots & \vdots \\ 
   \hline
\end{tabular}
\end{table}
After this step, we will have the eQTL gene expression files:
\begin{itemize}
    \item Skin\_Sun\_Exposed\_Lower\_leg\_gene\_expression.txt
    \item Whole\_Blood\_gene\_expression.txt
\end{itemize} 
  

 
## Preparing GWAS data
TisCoMM can handle two different types of GWAS input dataset, the individual level data and summary statistics. There are some difference between these two types of input, here we discuss them separately.


### Formatting GWAS individual data
The GWAS individual data files consist of genotype and phenotype data for all GWAS samples. They should be in the PLINK binary format. For example:
\begin{itemize}
    \item NFBC1966.bed,
    \item NFBC1966.bim,
    \item NFBC1966.fam.
\end{itemize}

You could optionally add the covariate file, which contains all confounding covariates used to adjust population stratification in the GWAS data. The covariate file should be formatted in a similar manner to the plink phenotype file, which should also be a \textbf{tab-delimited} file. 

### Formatting GWAS summary statisitc data

1. Reference panel in the PLINK binary format (.\textbf{bed}, .\textbf{bim}, .\textbf{fam}).
 For example,
 \begin{itemize}
     \item 1000G.bed,
     \item 1000G.bim,
     \item 1000G.fam.
 \end{itemize}
 
2. GWAS summary statistic data is required to have specific columns. See Table \ref{tab:gwasSS} for a demonstration. Note that all the column names should be exactly the same as those in Table \ref{tab:gwasSS}.If GWAS summary statistic in the original downloaded file do not come with all the information \textit{TisCoMM} needs, one will have to compute them manually. For example, if odds ratio is included, then beta can be computed as $\log(\text{Odds Ratio})$. Assume our interested trait is the late-onset Alzheimer's disease (LOAD), and we download the summary statistic file from \href{http://web.pasteur-lille.fr/en/recherche/u744/igap/igap_download.php}{http://web.pasteur-lille.fr/en/recherche/u744/igap/igap\_download.php}. After this step, the summary statistic is formatted correctly in following file:
\begin{itemize}
    \item LOAD.txt
\end{itemize}

\begin{table}[ht]
\centering
\caption{An example for the GWAS summary statistics.}
\label{tab:gwasSS}
\begin{tabular}{ccccccc}
  \hline
  SNP & chr & BP & A1 & A2 & beta & se \\ 
  \hline
rs3094315 &     1 & 752566 & G & A & -0.0122 & 0.0294 \\ 
rs3128117 &     1 & 944564 & C & T & -0.0208 & 0.0278 \\ 
rs1891906 &     1 & 950243 & C & A & -0.0264 & 0.0260 \\ 
rs2710888 &     1 & 959842 & T & C & -0.0439 & 0.0297 \\ 
rs4970393 &     1 & 962606 & G & A & -0.0252 & 0.0233 \\ 
rs7526076 &     1 & 998395 & A & G & -0.0512 & 0.0229 \\ 
  \vdots & \vdots &   \vdots & \vdots & \vdots & \vdots & \vdots  \\ 
   \hline
\end{tabular}
\end{table}



 

## Fit TisCoMM for GWAS individual data  
```{r, fig.show='hold', eval=FALSE}
# eQTL genotype file
file1 <- "GTEx_qc"

# GWAS individual level data
file2 <- "NFBC1966_qc"

# eQTL gene expression files
file3 <- c("Skin_Sun_Exposed_Lower_leg_gene_expression.txt",
           "Whole_Blood_gene_expression.txt")

# eQTL covariates file. Since normalized GE is provided, we do not need this file.
file4 <- ""

# GWAS covariates file
file5 <- ""

wihchPheno <- 1
bw         <- 5e5
coreNum    <- 24

fit <-mammot_paral(file1, file2, file3, file4, file5, 
                   wihchPheno, bw, coreNum)
```

There are other three arguments.

- whichPheno specifies which phenotype in the phenotype file (GTEX\_qc.fam) is used for association tests. 

- bw defines the cisSNPs within a gene: either up to bw proximal to the start of gene, or up to bw distal to the end of the gene.
    
- corNum sets the number of threads the program will use. 

## Fit TisCoMM for GWAS summary statistics

```{r, fig.show='hold', eval=FALSE}
# eQTL genotype file
file1 <- "GTEx_qc"

# GWAS summary statistic file
file2 <- "LOAD.txt"

# reference panel file
file3 <- "1000G"

# eQTL gene expression files
file4 <- c("Skin_Sun_Exposed_Lower_leg_gene_expression.txt",
           "Whole_Blood_gene_expression.txt")


# eQTL covariates file. Since normalized GE is provided, we do not need this file.
file5 <- ""

lam     <- 0.95
bw      <- 5e5
coreNum <- 24

 

fit <- mammotSS_paral(file1, file2, file3, file4, file5, 
                      lam, bw, coreNum)

```
There are other three arguments.

- lam is the shrinkage intensify for the reference panel. 

- bw defines the cisSNPs within a gene: either up to bw proximal to the start of gene, or up to bw distal to the end of the gene.

- corNum sets the number of threads the program will use. 



# Replicate simulation results in Shi et al. (2019)
All the simulation results can be reproduced by using the code at simulation. Before running simulation to reproduce the results, please familarize yourself with **TisCoMM** using 'TisCoMM' vignette. 

1. Simulation results for multi-tissue joint can be reproduced by following steps:

    - ExampleOne.R: This function can be run in a HPC cluster (with minor revisions, it could be run on a PC), it will output files, named pvalue_hz0.1_hc0.25_rhoX5_s5_batch-6.txt, which contain inference results of each replicate, for all multi-tissue TWAS  methods: TisCoMM, TisCoMM-S$^2$, MultiXcan, S-MultiXcan and UTMOST. 


    - ExampleOnePlot.R: This function produces simulation figures of joint test in Shi et al. (2019).

2. Simulation results for tissue-specific test can be reproduced by following steps:

     - PartCoMMCOR.R: This function can be run in a HPC cluster (with minor revisions, it could be run on a PC), it will output files, named part_hc4_rhoX8_rhoW8nz_ti2_batch-2.rds, which contain inference results of each replicate, for all single-tissue TWAS methods: CoMM, PrediXcan, and TWAS. 
     
    - SummaryCOR.R: This function produces simulation figures of tissue-specific test in Shi et al. (2019).

 
## Reference
[Shi et al (2019). A tissue-specific collaborative mixed model for jointly analyzing multiple tissues in transcriptome-wide association studies](https://www.biorxiv.org/content/10.1101/789396v3.full)
