# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

mammot_PXem <- function(x1, y, x2, z, w, constr, PX, maxIter) {
    .Call(`_mammot_mammot_PXem`, x1, y, x2, z, w, constr, PX, maxIter)
}

ADW <- function(x, y) {
    .Call(`_mammot_ADW`, x, y)
}

mammot_part_test <- function(x1, y, x2, z, w, constrFactor, PX, maxIter) {
    .Call(`_mammot_mammot_part_test`, x1, y, x2, z, w, constrFactor, PX, maxIter)
}

mammotSS_part_test <- function(x1, y, x2, z, w, constrFactor, PX, maxIter) {
    .Call(`_mammot_mammotSS_part_test`, x1, y, x2, z, w, constrFactor, PX, maxIter)
}

mammot_PXem_part <- function(x1, y, x2, z, w, constrFactor, PX, maxIter) {
    .Call(`_mammot_mammot_PXem_part`, x1, y, x2, z, w, constrFactor, PX, maxIter)
}

mammotSS_part_est <- function(x1, y, x2, z, w, B_hat, mu, Sigma, sigmab, Ve, alpha, constrFactor, PX, maxIter) {
    .Call(`_mammot_mammotSS_part_est`, x1, y, x2, z, w, B_hat, mu, Sigma, sigmab, Ve, alpha, constrFactor, PX, maxIter)
}

mammotSS_PXem_part <- function(x1, y, x2, z, w, constrFactor, PX, maxIter) {
    .Call(`_mammot_mammotSS_PXem_part`, x1, y, x2, z, w, constrFactor, PX, maxIter)
}

lmm_pxem <- function(y, w, x, maxIter) {
    .Call(`_mammot_lmm_pxem`, y, w, x, maxIter)
}

mammot_PXem_ss <- function(x1, y, x2, z, w, constr, PX, maxIter) {
    .Call(`_mammot_mammot_PXem_ss`, x1, y, x2, z, w, constr, PX, maxIter)
}

getColNum_Header <- function(filename, delimiter) {
    .Call(`_mammot_getColNum_Header`, filename, delimiter)
}

dataLoader <- function(stringname1, stringname2, stringname3, stringname4, stringname5, whCol) {
    .Call(`_mammot_dataLoader`, stringname1, stringname2, stringname3, stringname4, stringname5, whCol)
}

dataLoaderSS <- function(stringname1, stringname2, stringname3, stringname4, stringnameSS) {
    .Call(`_mammot_dataLoaderSS`, stringname1, stringname2, stringname3, stringname4, stringnameSS)
}

#' @title
#' mammot
#' @description
#'  
#'
#' @param stringname1  prefix for eQTL genotype file with plink format (bim, bed).
#' @param stringname2  prefix for reference panal GWAS genotype file with plink format (bim, bed, fam).
#' @param stringname3  gene expression file with full name.
#' @param stringname4  covariates file for eQTL data.
#' @param stringname5  GWAS summary statisitcs, which has specific form.
#' @param whCol  specify which phenotype is used in fam. For example, when whCol = 2, the seven-th column of fam file will be used as phenotype.
#' @param bw  the number of downstream and upstream SNPs that are considered as cis-SNP within a gene.
#'
#' @return List of model parameters
#'
#' @examples
#' ##Working with no summary statistics, no covariates and options
#' file1 = "1000G.EUR.QC.1";
#' file2 = "NFBC_filter_mph10";
#' file3 = "Geuvadis_gene_expression_qn.txt";
#' file4 = "";
#' file5 = "pc5_NFBC_filter_mph10.txt";
#' whichPheno = 1;
#' bw = 500000;
#'
#' fm = mammotSS_testing_run(file1,file2,file3, file4,file5, whichPheno, bw);
#'
#' @details
#' \code{mammotSS_testing_run} fits the mammot model. It requires to provide plink binary eQTL genotype file (bim, bed)
#' the GWAS summary statisitcs file (txt), gene expression file for eQTL.
#' @export
NULL

#' @title
#' CoMM
#' @description
#' CoMM to dissecting genetic contributions to complex traits by leveraging regulatory information.
#'
#' @param stringname1  prefix for eQTL genotype file with plink format (bim, bed).
#' @param stringname2  prefix for GWAS genotype and phenotype file with plink format (bim, bed, fam).
#' @param stringname3  gene expression file with full name.
#' @param stringname4  covariates file for eQTL data.
#' @param stringname5  covariates file for GWAS data, e.g. top 10 PCs.
#' @param whCol  specify which phenotype is used in fam. For example, when whCol = 2, the seven-th column of fam file will be used as phenotype.
#' @param bw  the number of downstream and upstream SNPs that are considered as cis-SNP within a gene.
#'
#' @return List of model parameters
#'
#' @examples
#' ##Working with no summary statistics, no covariates and options
#' file1 = "1000G.EUR.QC.1";
#' file2 = "NFBC_filter_mph10";
#' file3 = "Geuvadis_gene_expression_qn.txt";
#' file4 = "";
#' file5 = "pc5_NFBC_filter_mph10.txt";
#' whichPheno = 1;
#' bw = 500000;
#'
#' fm = CoMM_testing_run(file1,file2,file3, file4,file5, whichPheno, bw);
#'
#' @details
#' \code{CoMM} fits the CoMM model. It requires to provide plink binary eQTL genotype file (bim, bed)
#' the GWAS plink binary file (bim, bed, fam), gene expression file for eQTL.
#' @export
mammot <- function(stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw) {
    .Call(`_mammot_mammot`, stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw)
}

mammot_paral <- function(stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw, coreNum) {
    .Call(`_mammot_mammot_paral`, stringname1, stringname2, stringname3, stringname4, stringname5, whCol, bw, coreNum)
}

mammotSS <- function(stringname1, stringname2, stringname3, stringname4, stringname5, lam, bw) {
    .Call(`_mammot_mammotSS`, stringname1, stringname2, stringname3, stringname4, stringname5, lam, bw)
}

mammotSS_paral <- function(stringname1, stringname2, stringname3, stringname4, stringname5, lam, bw, coreNum) {
    .Call(`_mammot_mammotSS_paral`, stringname1, stringname2, stringname3, stringname4, stringname5, lam, bw, coreNum)
}

mammot_part_paral <- function(stringname1, stringname2, stringname3, stringname4, stringname5, targetList, whCol, bw, coreNum) {
    .Call(`_mammot_mammot_part_paral`, stringname1, stringname2, stringname3, stringname4, stringname5, targetList, whCol, bw, coreNum)
}

mammotSS_part_paral <- function(stringname1, stringname2, stringname3, stringname4, stringname5, targetList, lam, bw, coreNum) {
    .Call(`_mammot_mammotSS_part_paral`, stringname1, stringname2, stringname3, stringname4, stringname5, targetList, lam, bw, coreNum)
}

read_GWAS <- function(filename, P) {
    .Call(`_mammot_read_GWAS`, filename, P)
}

